### PR TITLE
Integrate error persistence into the projection process

### PR DESCRIPTION
**Pull request by olivergrabinski**
#3281

### CODE REVIEW

File: delta/plugins/elasticsearch/src/test/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/indexing/ElasticSearchCoordinatorSuite.scala [hunk 1]
```diff
@@ -190,11 +191,43 @@ class ElasticSearchCoordinatorSuite extends BioSuite with SupervisorSetup.Fixtur
...
              !createdIndices.contains(view3.index),
-             s"The index for '${view2.ref.viewId}' should not have been created."
+             s"The index for '${view3.ref.viewId}' should not have been created."
            )
     } yield ()
   }
 
+  test("There is one error for the coordinator projection before the signal") {
+    for {
+      entries <- projectionStore.failedElemEntries(ElasticSearchCoordinator.metadata.name, Offset.At(1L)).compile.toList
```
Comment (imsdu) [inline delta/plugins/elasticsearch/src/test/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/indexing/ElasticSearchCoordinatorSuite.scala:201]: You can use `Offset.start` here

File: delta/plugins/elasticsearch/src/test/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/indexing/ElasticSearchCoordinatorSuite.scala [hunk 2]
```diff
@@ -190,11 +191,43 @@ class ElasticSearchCoordinatorSuite extends BioSuite with SupervisorSetup.Fixtur
...
+      r        = entries.assertOneElem
+      _        = assertEquals(r.failedElemData.id, "https://bluebrain.github.io/nexus/vocabulary/view3")
+    } yield ()
+  }
+
+  test("There is one error for view 1") {
+    for {
+      entries <- projectionStore.failedElemEntries(view1.projection, Offset.At(1L)).compile.toList
```
Comment (imsdu) [inline delta/plugins/elasticsearch/src/test/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/indexing/ElasticSearchCoordinatorSuite.scala:209]: Same

File: delta/sourcing-psql/src/main/scala/ch/epfl/bluebrain/nexus/delta/sourcing/stream/Projection.scala [hunk 3]
```diff
@@ -57,35 +59,61 @@ final class Projection private[stream] (
...
+                         case ExitCase.Error(th) => status.update(_.failed(th))
+                         case ExitCase.Completed => Task.unit // streams stopped through a signal still finish as Completed
+                         case ExitCase.Canceled  => Task.unit // the status is updated by the logic that cancels the stream
+                       }
+                       .mapAccumulate(progress) {
+                         case (acc, msg) if msg.offset.value > progress.offset.value => (acc + msg, msg)
+                         case (acc, msg)                                             => (acc, msg)
+                       }
+                       .groupWithin(batch.maxElements, batch.maxInterval)
+                       .map { elements =>
```
Comment (imsdu) [inline delta/sourcing-psql/src/main/scala/ch/epfl/bluebrain/nexus/delta/sourcing/stream/Projection.scala:89]: * Join the `map` and the `evalTap` in a single `evalTap`
* Extract the content in a function outside the for comprehension

File: delta/sourcing-psql/src/main/scala/ch/epfl/bluebrain/nexus/delta/sourcing/stream/Projection.scala [hunk 4]
```diff
@@ -57,35 +59,61 @@ final class Projection private[stream] (
...
+                           case ((_, failedElems), (newProgress, _))                =>
+                             (Some(newProgress), failedElems)
+                         }
+                       }
+                       .evalTap {
+                         case (None, _)                        => UIO.unit
+                         case (Some(newProgress), failedElems) =>
+                           UIO.delay(
+                             logger.debug(
+                               s"[${projection.metadata.name}] Progress: ${newProgress.offset.value}. ${failedElems.length} failed elems to save."
```
Comment (imsdu) [inline delta/sourcing-psql/src/main/scala/ch/epfl/bluebrain/nexus/delta/sourcing/stream/Projection.scala:102]: For debug logs, it is better to use parameterized messages:
https://www.slf4j.org/faq.html#logging_performance

File: delta/sourcing-psql/src/main/scala/ch/epfl/bluebrain/nexus/delta/sourcing/stream/Projection.scala [hunk 5]
```diff
@@ -57,35 +59,61 @@ final class Projection private[stream] (
...
+                       .map { elements =>
+                         elements.foldLeft((Option.empty[ProjectionProgress], List.empty[FailedElem])) {
+                           case ((_, failedElems), (newProgress, elem: FailedElem)) =>
+                             (Some(newProgress), failedElems :+ elem)
+                           case ((_, failedElems), (newProgress, _))                =>
+                             (Some(newProgress), failedElems)
+                         }
+                       }
+                       .evalTap {
+                         case (None, _)                        => UIO.unit
```
Comment (imsdu) [inline delta/sourcing-psql/src/main/scala/ch/epfl/bluebrain/nexus/delta/sourcing/stream/Projection.scala:98]: You can replace the pattern matching by something like
```scala
newProgress.traverse { np =>  progressRef.set(np) >> np(newProgress)} 
>>
UIO.when(failedElems.nonEmpty)(saveFailedElems(failedElems))
```

File: delta/sourcing-psql/src/main/scala/ch/epfl/bluebrain/nexus/delta/sourcing/stream/ProjectionStore.scala [hunk 6]
```diff
@@ -185,12 +187,11 @@ object ProjectionStore {
...
           metadata: ProjectionMetadata,
           failures: List[FailedElem]
-      ): UIO[Unit] =
-        failures
-          .traverse(elem => saveFailedElem(metadata, elem))
-          .transact(xas.write)
-          .void
-          .hideErrors
+      ): UIO[Unit] = {
+        val log  = UIO(logger.debug(s"[${metadata.name}] Saving ${failures.length} failed elems."))
```
Comment (imsdu) [inline delta/sourcing-psql/src/main/scala/ch/epfl/bluebrain/nexus/delta/sourcing/stream/ProjectionStore.scala:191]: Seems to only repeat the previous one

**Review by imsdu — CHANGES_REQUESTED** [review]

File: delta/sourcing-psql/src/main/scala/ch/epfl/bluebrain/nexus/delta/sourcing/stream/ProjectionStore.scala (hunk 6 above)
Comment (olivergrabinski) [reply delta/sourcing-psql/src/main/scala/ch/epfl/bluebrain/nexus/delta/sourcing/stream/ProjectionStore.scala]: I kept the one inside the `saveFailedElems` function.

**Review by imsdu — APPROVED** [review]

**Review by imsdu — APPROVED** [review]

**Pull request merged**